### PR TITLE
Deprecate wildcard assignments.

### LINKED ISSUE #3314: Less error prone tuple assignments
(This has been also mentioned by the recent Solidity audit under SOL-007.)

Currently tuple assignments can skip leading, trailing and positional variable names.  I'd propose to make these rules strict and require the user to specify the required number of return values by appropriate number of variable names. If omitted either accept an empty name or require `_`.

e.g. `(,a,,,,b,,,,,c,,,,,) = crazyfunction()` or `(_,a,_,_,_,_b,_,_,_,_,_c,_,_,_,_,_) = crazyfunction()`

If such a construct is deemed unreadable by the user that should trigger a refactoring of the code to avoid returning countless return values, which probably will benefit the code quality.

(`_` makes sense because some languages (like Lua) already do that, but it also conflicts with the modifier body marker `_`)

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/3314

TODO:
 - [x] fix special case of `(uint x) = (1,)` (i.e. count the number of components in a tuple expression as one if there are exactly two and the second is empty)
 - [x] adjust tests
 - [x] document
 - [x] changelog entry

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1345,11 +1368,30 @@ bool TypeChecker::visit(Assignment const& _assignment)
...
 		_assignment.annotation().type = make_shared<TupleType>();
+
 		expectType(_assignment.rightHandSide(), *tupleType);
 
 		// expectType does not cause fatal errors, so we have to check again here.
-		if (dynamic_cast<TupleType const*>(type(_assignment.rightHandSide()).get()))
+		if (TupleType const* rhsType = dynamic_cast<TupleType const*>(type(_assignment.rightHandSide()).get()))
+		{
 			checkDoubleStorageAssignment(_assignment);
+			// TODO For 0.5.0, this code shoud move to TupleType::isImplicitlyConvertibleTo,
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: I think we were using `@todo` in other place (not sure why), but keeping the same makes it easier to find these.

**Comment by axic** [comment]
Looks good.

**Comment by chriseth** [comment]
It turns out the special case is not necessary, since the type (singleton tuple) is already properly assigned in the type checker.

**Comment by chriseth** [comment]
@axic ready for review.

**Comment by axic** [comment]
Rebased.

As mentioned before I was happy with the PR as is, but I think found a couple of questions now.

File: test/libsolidity/syntaxTests/tupleAssignments/nowarn_explicit_singleton_token_expression.sol [hunk 2]
```diff
@@ -0,0 +1,8 @@
+contract C {
+	function f() public pure {
+        uint a;
+		(a,) = (uint(1),);
+	}
+}
+// ----
+// Warning: (59-76): Different number of components on the left hand side (2) than on the right hand side (1).
```
Comment (axic) [inline test/libsolidity/syntaxTests/tupleAssignments/nowarn_explicit_singleton_token_expression.sol]: Why is the RHS accepted, but being told it is 1 component?

**Comment by axic** [comment]
Fixed indentation issues and added a new test `large_component_count.sol`.

**Review by axic — APPROVED** [review]
Apart from the that single question (which probably raises a problem which needs to be discussed separate).

File: test/libsolidity/syntaxTests/tupleAssignments/nowarn_explicit_singleton_token_expression.sol (hunk 2 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/tupleAssignments/nowarn_explicit_singleton_token_expression.sol]: This is the way to distinguish between a parenthesized expression and a tuple with one component: https://solidity.readthedocs.io/en/v0.4.23/control-structures.html#destructuring-assignments-and-returning-multiple-values (in the code)

**Pull request merged**